### PR TITLE
feat(cdp): add cdp destination APIs

### PR DESCRIPTION
**Pull request by hazzadous**
This commit adds the CDP destination APIs. Key changes include:

 - use `sqlx` for migrations
 - jest for functional_tests (although I would be happy to use vitest or
   alternatives if we want to, I didn't want to change too much at once)
 - pnpm for package management
 - koajs for the server
 - JWT for authn/authz. We can reuse some of the stuff @benjackwhite has put in place for the exports.
 - Ajv for validation
 - A separate PostgreSQL logical database for the destination APIs
   persistence.

Things still to do:

 - add some delivery mechanism that takes events from Kafka and puts
   them to the destinations.
 - add CI
 - add to Helm Chart
 - add some method of authentication. I've added the API here but it
   might be that I just end up putting that in the main app in the end,
   depending on how much momentum there is to try out separating the API
   a bit, and the logistics of that.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: cdp/src/sql-template-string.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export const SQL = (sqlParts: TemplateStringsArray, ...args: any[]): { text: string; values: any[] } => {
```
Comment (lharries) [inline cdp/src/sql-template-string.ts:1]: This is cool! We just have to not forget the SQL before the backticks

File: cdp/src/destination-types/handlers.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    description: string
+    configSchema: Record<string, unknown> // A JSONSchema describing the configuration
+}
+
+const destinationTypes: { [type: string]: DestinationType } = {
+    webhook: {
+        type: 'webhook',
+        name: 'Webhook',
+        description: 'Send events to a webhook',
+        configSchema: {
```
Comment (lharries) [inline cdp/src/destination-types/handlers.ts:22]: further down the line we'll want to add the ability to set headers too

File: openapi/customer-data-pipeline.yaml [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                        type: string
+                        description: Name of the destination
+                    description:
+                        type: string
+                        description: Description of the destination
+                    created_at:
+                        type: string
+                        format: date-time
+                        description: Date and time when the destination was created
+                    updated_at:
```
Comment (lharries) [inline openapi/customer-data-pipeline.yaml:363]: I think this is missing the `configSchema`

File: openapi/customer-data-pipeline.yaml [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+                    content:
+                        application/json:
+                            schema:
+                                $ref: '#/components/schemas/Error'
+
+    components:
+        schemas:
+            DestinationType:
+                type: object
+                properties:
```
Comment (lharries) [inline openapi/customer-data-pipeline.yaml:326]: I think this is missing `configSchema`

**Comment by lharries** [comment]
Good work so far, added some comments but all seems sensible to me :)

**Comment by lharries** [comment]
I've gone through what I think the rough UI/UX for this should be ([interactive figma here](https://www.figma.com/proto/XCs0inErjInoCNo1Xz8jj0/CDP-UI?node-id=2-126&scaling=scale-down&page-id=1%3A20&starting-point-node-id=2%3A126&show-proto-sidebar=1)) or [raw designs here](https://www.figma.com/file/XCs0inErjInoCNo1Xz8jj0/CDP-UI?node-id=1-20&t=e0LhcqMG5QfMkeHj-0)

A few considerations that came up when doing this:
- Should each "destination" mean one set of (filters, transformation, url, headers) or should we instead do what Segment does and have the destination be a group of "mappings" where each mapping = (filters, transformation, url, headers) and then the destination has the name, description etc.? (For reference, Rudderstack does the first option in contrast to Segment)
  - **I'd propose that we go with the second option where each destination has 1+ mappings**. The reason for this is that when we later add 1st party destinations nearly all of these will have multiple mappings, and so as I don't think it's much more complex we should likely do this from the start. It also means that they won't have 4 destinations just for customer.io, but instead just 1 destination which I think is semantically more meaningful and will reduce bloat.
- Should we/How do we want to enable transformations? 
  - I've looked at a few options here - we could not bother, we could do a custom basic templating system similar to f-strings, we could use a general purpose templating library like jinja, we could write all the transformations in javascript or python (and enable our users to do that too), or we could use a JSON specific templating library.
  - I've looked into the JSON specific templating libraries further and JSONata looks the most promising. (I also evaluated JQ, JSONet, and JSLT).
  - **I'd propose we go with JSONata**. Why? We'll need something expressive enough such that later we can add lots of the transformations. The declarative language will save us lots of boilerplate code. It is simple and quick to get started. It's written in pure javascript so easy to implement. It has 1.5k stars so it's reasonably popular: https://snyk.io/advisor/npm-package/jsonata It also looks quick to implement so worth doing it
  - You can try it out [here](https://try.jsonata.org/). Here's an example
  Screenshot JSONata Exerciser (Google Chrome) 2023-04-11 at 19 58@2x


It would be great if you could go through the designs, particularly of the mapping part and check that this is what you have in mind too

File: cdp/src/destination-types/handlers.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            type: 'object',
+            properties: {
+                url: {
+                    type: 'string',
+                    description: 'The URL to send the webhook to',
+                },
+            },
+            required: ['url'],
```
Comment (lharries) [inline cdp/src/destination-types/handlers.ts:30]: As discussed here: https://github.com/PostHog/posthog/pull/14994#issuecomment-1504359549 I think instead we should have each destination can have many configs likely in their own table (but will let you decide on the implementation)

File: cdp/src/sql-template-string.ts (hunk 1 above)
Comment (hazzadous) [reply cdp/src/sql-template-string.ts:1]: Good point! Will add some defence against this 😱

File: cdp/src/destination-types/handlers.ts (hunk 2 above)
Comment (hazzadous) [reply cdp/src/destination-types/handlers.ts:22]: Yep I haven't filled out much on this part yet. I'm going to go deep first and get the stack working end to end e.g.:

 1. get the basic API deployed
 2. get the frontend able to call it

Then we can do follow-ups to add the requirements here.

**Comment by hazzadous** [comment]
>   * **I'd propose that we go with the second option where each destination has 1+ mappings**. The reason for this is that when we later add 1st party destinations nearly all of these will have multiple mappings, and so as I don't think it's much more complex we should likely do this from the start. It also means that they won't have 4 destinations just for customer.io, but instead just 1 destination which I think is semantically more meaningful and will reduce bloat.

Yes makes sense. Will update the docs accordingly.

>   * **I'd propose we go with JSONata**. Why? We'll need something expressive enough such that later we can add lots of the transformations. The declarative language will save us lots of boilerplate code. It is simple and quick to get started. It's written in pure javascript so easy to implement. It has 1.5k stars so it's reasonably popular: https://snyk.io/advisor/npm-package/jsonata It also looks quick to implement so worth doing it

Let's do it!

**Comment by hazzadous** [comment]
I'll yolo this in then make the requested changes in follow-up PRs 🙏

**Pull request merged**